src/main/java/fi/aalto/cs/apluscourses/intellij/utils/Interfaces.java

```diff
@@ -88,6 +114,133 @@ public static void refreshPath(VirtualFile file, Runnable callback) {
+      try {
+        byte[] hash = digest.digest(Files.readAllBytes(path));
+        return Base64.getEncoder().encodeToString(hash);
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+    }
+
+    /**
+     * Checks whether the same submission has already been submitted by the user. If it is the case, the user
+     * is asked whether they want to submit anyway or abort the submission.
+     * @param project The current IntelliJ project.
+     * @param courseId ID of the course which contains the exercise in question.
+     * @param exerciseId ID of the exercise.
+     * @param files A map of submittable files, as constructed in {@link SubmitExerciseAction}.
+     * @return True if duplicate check succeeded and the submission should proceed; false if it should be cancelled.
+     */
+    public static boolean checkForDuplicateSubmission(@NotNull Project project,
+                                                      @NotNull String courseId,
+                                                      long exerciseId,
+                                                      @NotNull Map<String, Path> files) {
+      // we want to defer creation of the cache until it's actually necessary
+      if (submissionsCache == null) {
+        Path projectPath = Path.of(Objects.requireNonNull(project.getBasePath()));
+        submissionsCache = new JsonFileCache(
+            projectPath.resolve(Path.of(Project.DIRECTORY_STORE_FOLDER, "a-plus-hashes.json")));
+      }
+
+      try {
+        final MessageDigest shaDigest = MessageDigest.getInstance("SHA-256");
+        final StringBuilder submissionString = new StringBuilder();
+
+        // the iteration order is important so that files always get concatenated in the same order
+        files.entrySet().stream().sorted(Map.Entry.comparingByKey()).forEachOrdered(entry -> {
+          submissionString.append(entry.getKey());
+          submissionString.append('|'); // a separating character that is not in base-64 alphabet
+          submissionString.append(hashFileToString(shaDigest, entry.getValue()));
+          submissionString.append(',');
+        });
+
+        final String cacheKey = "hash_c" + courseId + "_e" + exerciseId;
+        final byte[] finalHashBytes = shaDigest.digest(submissionString.toString().getBytes(StandardCharsets.UTF_8));
+        final String finalHash = Base64.getEncoder().encodeToString(finalHashBytes);
+
+        final JSONObject cachedHashesJson = submissionsCache.getValue(cacheKey, CachePreferences.PERMANENT);
+        final SubmissionHashes cachedHashes = new SubmissionHashes(cachedHashesJson);
+        if (cachedHashesJson == null) {
+          // this is the first hash, so the entry doesn't yet exist in the cache
+          cachedHashes.addHash(finalHash);
+          submissionsCache.putValue(cacheKey, cachedHashes.toJsonObject(), CachePreferences.PERMANENT);
+          return true;
+        }
```

Comment:
I don't see reason for this conditional block. It could be removed and the code would function the same, right?

Reply:
Correct, this conditional block is a remnant of a past version that used different caching method